### PR TITLE
fixed the build issues

### PR DESCRIPTION
**Pull request by jimmymannekkattu**
### Checklist:

- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

### CODE REVIEW

File: plugins/linux-v4l2/v4l2-decoder.c
Comment (jimmymannekkattu) [inline plugins/linux-v4l2/v4l2-decoder.c:1]: avcodec_close is depricated. avcodec_free_context can be used instead.

File: plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c
Comment (jimmymannekkattu) [inline plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c:1]: casting the function pointer with right signature.

File: plugins/obs-ffmpeg/obs-ffmpeg-mpegts.c
Comment (jimmymannekkattu) [inline plugins/obs-ffmpeg/obs-ffmpeg-mpegts.c:1]: casting the function pointer with right signature.

**Comment by Fenrirthviti** [comment]
Please properly fill out the template, and fix your commits according to our contribution guidelines.

File: plugins/linux-v4l2/v4l2-decoder.c
Comment (jimmymannekkattu) [inline plugins/linux-v4l2/v4l2-decoder.c:1]: Reference Here -> https://ffmpeg.org/doxygen/trunk/group__lavc__core.html#gab7f632082a10fcb051fa3e2db7f86942

File: plugins/obs-ffmpeg/obs-ffmpeg-mpegts.c [hunk 1]
```diff
@@ -437,11 +437,11 @@ static inline int allocate_custom_aviocontext(struct ffmpeg_output *stream,
 	if (is_rist)
 		s = avio_alloc_context(
 			buffer, buffer_size, AVIO_FLAG_WRITE, h, NULL,
-			(int (*)(void *, uint8_t *, int))librist_write, NULL);
+			(int (*)(void *, const uint8_t *, int))libsrt_write, NULL);
```
Comment (RytoEX) [inline plugins/obs-ffmpeg/obs-ffmpeg-mpegts.c:440]: You changed `librist_write` to `libsrt_write` in the conditional block for `is_rist`. Revert that change.

File: plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c [hunk 2]
```diff
@@ -975,7 +975,7 @@ static inline int open_output_file(struct ffmpeg_mux *ffm)
 
 			ffm->output->pb = avio_alloc_context(
 				avio_ctx_buffer, AVIO_BUFFER_SIZE, 1, ffm, NULL,
-				ffmpeg_mux_write_av_buffer,
+				 (int (*)(void *, const uint8_t *, int))ffmpeg_mux_write_av_buffer,
```
Comment (RytoEX) [inline plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c:978]:
```suggestion
				(int (*)(void *, const uint8_t *, int))ffmpeg_mux_write_av_buffer,
```

File: plugins/linux-v4l2/v4l2-decoder.c [hunk 3]
```diff
@@ -78,7 +78,7 @@ void v4l2_destroy_decoder(struct v4l2_decoder *decoder)
 	}
 
 	if (decoder->context) {
-		avcodec_close(decoder->context);
+		// avcodec_close(decoder->context);
```
Comment (RytoEX) [inline plugins/linux-v4l2/v4l2-decoder.c:81]: Don't comment out code. Just remove it.

File: plugins/linux-v4l2/v4l2-decoder.c [hunk 4]
```diff
@@ -78,7 +78,7 @@ void v4l2_destroy_decoder(struct v4l2_decoder *decoder)
...
 		avcodec_free_context(&decoder->context);
```
Comment (RytoEX) [inline plugins/linux-v4l2/v4l2-decoder.c:82]: In what version of FFmpeg does this break? How far back is this syntax permitted/recommended? We have to support as far back as FFmpeg 4.4.2 due to still supporting Ubuntu 22.04.

**Review by RytoEX — CHANGES_REQUESTED** [review]
Please fill out the PR template properly. Not filling out the PR template may result in your PR being closed.

Please do not use merge commits. Use `git rebase` to keep up with changes to our master branch.

File: plugins/linux-v4l2/v4l2-decoder.c (hunk 4 above)
Comment (jimmymannekkattu) [reply plugins/linux-v4l2/v4l2-decoder.c:82]: `
#define FF_API_AVCODEC_CLOSE       (LIBAVCODEC_VERSION_MAJOR < 62)
#if FF_API_AVCODEC_CLOSE
/**
 * Close a given AVCodecContext and free all the data associated with it
 * (but not the AVCodecContext itself).
 *
 * Calling this function on an AVCodecContext that hasn't been opened will free
 * the codec-specific data allocated in avcodec_alloc_context3() with a non-NULL
 * codec. Subsequent calls will do nothing.
 *
 * @deprecated Do not use this function. Use avcodec_free_context() to destroy a
 * codec context (either open or closed). Opening and closing a codec context
 * multiple times is not supported anymore -- use multiple codec contexts
 * instead.
 */
attribute_deprecated
int avcodec_close(AVCodecContext *avctx);
#endif`

**Comment by Fenrirthviti** [comment]
Since this appears to be fixing a build issue that is outside anything we actually support, and despite repeated requests the template has not been properly filled out and context provided for what this is actually trying to solve, I'm going to close this as it seems like a PR to fix a local build issue, not a project build issue. Our internal CI and local builds are working just fine.

File: plugins/linux-v4l2/v4l2-decoder.c (hunk 4 above)
Comment (jimmymannekkattu) [reply plugins/linux-v4l2/v4l2-decoder.c:82]: https://ffmpeg.org/doxygen/trunk/group__lavc__core.html#gab7f632082a10fcb051fa3e2db7f86942
Reference is given above.

**Comment by jpetazzo** [comment]
When rebuilding OBS on my Archlinux system today, I ran into a typing issue - in [plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c](https://github.com/obsproject/obs-studio/pull/10358/files/e058ef83f231ec48909f9d37518f3e1e031a8eb7#diff-2bc93e6134b6ecd1adb0405b942d41f65adf5b7e57da02845ab3b7e3eac4116d) line 978, `ffmpeg_mux_write_av_buffer` should be of type `int (*)(void *, const uint8_t *, int)` but instead is of type `int (*)(void *, uint8_t *, int)`.

It _might_ be linked to changes in ffmpeg 7 (it _might_ be the case that I hadn't rebuilt OBS with ffmpeg 7 yet), but I'm not sure.

I appled a one-line patch to cast `ffmpeg_mux_write_av_buffer` and that allowed OBS to build successfully. [This](
https://github.com/obsproject/obs-studio/pull/10358/files/e058ef83f231ec48909f9d37518f3e1e031a8eb7#diff-2bc93e6134b6ecd1adb0405b942d41f65adf5b7e57da02845ab3b7e3eac4116dL978) is the corresponding change.

I don't know if the problem is within OBS source code, or something weird on my machine, or the phases of the moon; but my search led me to this GitHub issue, so I'm leaving a note here just in case others run into the same problem. 

Cheers!

**Comment by jimmythomas** [comment]
This was the fix provided for the same but the reviewer were looking on the build server as it was building with old version.

**Pull request closed without merging**